highlighters/main/main-highlighter.zsh

```diff
@@ -128,22 +128,36 @@ _zsh_highlight_main_calculate_fallback() {
 # Uses the zsh/parameter module if available to avoid forks, and a
 # wrapper around 'type -w' as fallback.
 #
-# Takes a single argument.
+# If $2 is 0, do not consider aliases.
 #
 # The result will be stored in REPLY.
 _zsh_highlight_main__type() {
+  integer -r aliases_allowed=${2-1}
+  # We won't cache replies of anything that exists as an
+  # alias at all, regardless of $aliases_allowed.
```

Comment:
Say why: because the __type of a word that exists as an alias depends on the value of $aliases_allowed.  (That's why the "We probably should" in the next sentence makes sense.)